Package best practices doc (#139)
dbt-labs/hubcap: package-best-practices.md

# Best practices when submitting a package for public use

> Packages are a key part of the dbt experience. Practitioners and vendors alike can [contribute to the knowledge loop](https://github.com/dbt-labs/corp/blob/main/values.md#we-contribute-to-the-knowledge-loop) by enabling compelling, batteries-included use cases.

Reviewer: 🎯

There are hundreds of packages on the dbt Package Hub already, and over time we have identified a handful of common mistakes that new package authors make as they take a self-contained package and prepare it for publication. We have listed these below to help you and your users get the best experience.

Although the dbt Labs team completes a cursory review of new packages before adding them to the Package Hub, ultimately you are responsible for your package's performance.

> To avoid ambiguity, the bullets below are pretty formal, but we are a friendly bunch! If you need help preparing a package for deployment to the dbt Package Hub, reach out in [#package-ecosystem](https://getdbt.slack.com/archives/CU4MRJ7QB/) in the [dbt Community Slack](https://getdbt.com/community).

Reviewer: 🙌

---

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

> ### First run experience
> - Packages SHOULD contain a README file which explains how to get started with the package and customise its behaviour.

Reviewer: ```suggestion
### First run experience
- Packages SHOULD contain a README file which explains how to get started with the package and customise its behaviour.
- The documentation SHOULD indicate which data warehouses are expected to work with this package.
```

### Customisability 
- Packages MUST NOT hard-code table references, and MUST use `ref` or `source` instead.
- Packages MAY provide a mechanism (such as variables) to enable/disable certain subsets of functionality.
- Packages for data transformation:
    - SHOULD provide a mechanism (such as variables) to customise the location of source tables.
    - SHOULD NOT assume database/schema names in sources. 
    - MAY assume table names, particularly if the package was built to support tables created by a known tool.

### Dependencies
- Packages requiring a minimum version of `dbt-core` MUST declare it in the `require-dbt-version` property of `dbt_project.yml`.
- Packages requiring a minimum version of `dbt-core` SHOULD declare compatibility with all subsequent minor and patch releases of that major version. 
    - For example, a package which depends on functionality added in dbt Core 1.2 SHOULD set its `require-dbt-version` property to `[">=1.2.0", "<2.0.0"]`.
Packages with a dependency on another package:

>     - MAY, if the dependent package is major-version-zero, pin to the current minor version only.

Author: Does "dependent" make clear that it is the package that the main package is importing? I think it does until I think too hard about dependents in the real world normally being children, which would make them the downstream one as opposed to looking upstream. Not sure if I'm too in my head or not, and don't have a better word.

Reviewer: Would "requirement" be a less confusing word?

An additional possibility: the "Dependencies" could be split into two parts:
- compatible versions of dbt Core
- compatible versions of package requirements

The first bullet point references `dbt_project.yml` -- maybe there could be one could referencing `packages.yml` also?

- Each package requirement MUST be declared in `packages.yml`.
- Requirement ranges SHOULD be as wide as compatibility allows.

☝️ might need some further word-smithing

    - SHOULD NOT pin to a patch version of the dependent package unless they are aware of an incompatibility between their package and the dependency.
    - SHOULD reference the version in the dbt Package Hub when available, as opposed to a `git` installation.

### Interoperability
- Packages MUST NOT override dbt Core behaviour in such a way as to impact other dbt resources (models, tests, etc) not provided by the package.
- Packages SHOULD disambiguate their resource names to avoid clashes with nodes that are likely to already exist in a project. 
    - For example, packages SHOULD NOT provide a model simply called `users`.

> ### Updates
> - Packages SHOULD follow the guidance of the [Semantic Versioning Specification](https://semver.org/).

Reviewer: ```suggestion
### Releases and updates
- Packages SHOULD follow the guidance of the [Semantic Versioning Specification](https://semver.org/).
- Packages MUST create a git tag with the version number that follows the accepted version format.
```
We'd probably need to be more clear about what the "accepted version format" is -- currently it is _strictly_ https://semver.org/spec/v2.0.0.html but would be relaxed slightly if #130 is merged.

    - SHOULD provide a mechanism (such as variables) to customise the location of source tables.
    - SHOULD NOT assume database/schema names in sources. 
    - MAY assume table names, particularly if the package was built to support tables created by a known tool.

### Dependencies
#### Dependencies on dbt Core
- Packages requiring a minimum version of `dbt-core` MUST declare it in the `require-dbt-version` property of `dbt_project.yml`.
- Packages requiring a minimum version of `dbt-core` SHOULD allow all subsequent minor and patch releases of that major version. 
    - For example, a package which depends on functionality added in dbt Core 1.2 SHOULD set its `require-dbt-version` property to `[">=1.2.0", "<2.0.0"]`.
#### Dependencies on other packages defined in `packages.yml`:
- Packages SHOULD import their dependencies from the dbt Package Hub when available, as opposed to a `git` installation.
- Packages SHOULD specify the widest possible range of supported versions, to minimise issues in dependency resolution. 
    - In particular, packages SHOULD NOT pin to a patch version of their imported package unless they are aware of an incompatibility.
### Interoperability

> - Packages MUST NOT override dbt Core behaviour in such a way as to impact other dbt resources (models, tests, etc) not provided by the package.

Author: @jtcohen6 I wrote this thinking about the aftermath of that Fivetran package that overrode freshness collection and had unfortunate side effects. 

But, if someone like [this community member](https://getdbt.slack.com/archives/C01NH3F2E05/p1659364136316949) turned their [gist](https://gist.github.com/LewisDavies/f126271284df804c3cf5cc9b094db6a8) into a package, that would be an acceptable overriding of behaviour yeah? Does this rule need a bit of wordsmithing or do we just drop it down to `REALLY SHOULD NOT` per [RFC 6919](https://datatracker.ietf.org/doc/html/rfc6919)?

Reviewer: +1 for `REALLY SHOULD NOT`

Author: `REALLY SHOULD NOT` doesn't actually exist 😢 (discussed above) - should've made clearer that this was a joke

Although the dbt Labs team completes a cursory review of new packages before adding them to the Package Hub, ultimately you are responsible for your package's performance.

To avoid ambiguity, the bullets below are pretty formal, but we are a friendly bunch! If you need help preparing a package for deployment to the dbt Package Hub, reach out in [#package-ecosystem](https://getdbt.slack.com/archives/CU4MRJ7QB/) in the [dbt Community Slack](https://getdbt.com/community).

---

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

### First run experience
- Packages SHOULD contain a README file which explains how to get started with the package and customise its behaviour.
- The documentation SHOULD indicate which data warehouses are expected to work with this package.

### Customisability 

> - Packages MUST NOT hard-code table references, and MUST use `ref` or `source` instead.

Reviewer: Non-blocking commentary.

The current text definitely conveys the _spirit_ of the guidance, but an overly lawyerly maintainer could hard-code a view reference and claim full compliance.

Two opportunities for further clarity:
- table → relation
- use applicable adapter methods rather than directly utilizing `describe`, `show`, or `information_schema`

I'm fuzzy on the detailed guidance to accomplish the latter, so that's why I don't have an explicit suggestion that covers both opportunities.

Our glossary doesn't include an entry "relation" currently, it only has the following disparate entries:
- https://docs.getdbt.com/terms/table
- https://docs.getdbt.com/terms/view
- https://docs.getdbt.com/terms/cte

The closest documentation I could find quickly focuses on the definition and usage of the Relation Python object:
- https://docs.getdbt.com/reference/dbt-classes#relation

- Packages MAY provide a mechanism (such as variables) to enable/disable certain subsets of functionality.
- Packages for data transformation:
    - SHOULD provide a mechanism (such as variables) to customise the location of source tables.
    - SHOULD NOT assume database/schema names in sources. 
    - MAY assume table names, particularly if the package was built to support tables created by a known tool.

### Dependencies
#### Dependencies on dbt Core
- Packages requiring a minimum version of `dbt-core` MUST declare it in the `require-dbt-version` property of `dbt_project.yml`.
- Packages requiring a minimum version of `dbt-core` SHOULD allow all subsequent minor and patch releases of that major version. 
    - For example, a package which depends on functionality added in dbt Core 1.2 SHOULD set its `require-dbt-version` property to `[">=1.2.0", "<2.0.0"]`.
#### Dependencies on other packages defined in `packages.yml`:
- Packages SHOULD import their dependencies from the dbt Package Hub when available, as opposed to a `git` installation.

> - Packages SHOULD NOT pin to a patch version of their imported package unless they are aware of an incompatibility.
> - Packages importing another package which has reached major version one or later SHOULD allow all subsequent minor and patch releases of that major version. 
> - Packages importing another package which is major version zero MAY pin to the current minor version only.

Reviewer: I think the last 3 bullet points can all be replaced with simple guidance:

```suggestion
- The widest version range possible SHOULD be specified to give the dependency resolver the appropriate latitude.
```

My guess is that after lots of discussion exploring many scenarios, ☝️ will still hold as an essential summary, a [cataclysm sentence](https://radiolab.org/episodes/cataclysm-sentence) that "contains the most information in the fewest words".

The three bullet points make assumptions about the versioning scheme being used by the imported package which _might_ be strict [SemVer 2.0](https://semver.org/spec/v2.0.0.html), but just as likely might be a variant or something else entirely.

Author: >The three bullet points make assumptions about the versioning scheme being used by the imported package which might be strict [SemVer 2.0](https://semver.org/spec/v2.0.0.html), but just as likely might be a variant or something else entirely.

Could we change that `SHOULD` from Releases and Updates to a `MUST`? 

Those bullets are written in the same spirit as when you go to a fancy hotel and there's a _No Elephants Allowed In the Ballroom_ sign - you know they didn't put it up until they had to. 

The first one (no patches) is the most common and catastrophic failure state I see; I'm not wed to the others but they seemed like the obvious next question: "If I'm not allowed to pin to a patch, what _do_ you recommend?"

What do you think of something like this?
> Packages SHOULD specify the widest possible range of supported versions, to minimise issues in dependency resolution. In particular, packages SHOULD NOT pin to a patch version of their imported package unless they are aware of an incompatibility.

Reviewer: > Packages SHOULD specify the widest possible range of supported versions, to minimise issues in dependency resolution. In particular, packages SHOULD NOT pin to a patch version of their imported package unless they are aware of an incompatibility.

🎯 looks perfect

> Could we change that `SHOULD` from Releases and Updates to a `MUST`?

I'm _guessing_ that we can't say "Packages MUST follow the guidance of Semantic Versioning 2.0.0" because we don't have any auditing mechanism -- there is no easy way to validate to what degree a package is following it or not. We _can_ validate the version string with a regex, but we _can't_ reasonably determine if they are following the SemVer guidance or not. They are on the honor system for this one.

Author: Fair!

### Interoperability
- Packages MUST NOT override dbt Core behaviour in such a way as to impact other dbt resources (models, tests, etc) not provided by the package.

> - Packages MUST use the cross-database macros built into dbt Core where available, such as `{{ datediff }}` and `{{ type_string() }}`.

Reviewer: `MUST` or `SHOULD`? Could a BigQuery-specific spend package BigQuery-isms? Or what if the current cross-database implementation doesn't cover some edge case that the package needs or wants to accommodate?

Optional ideas for re-wording slightly:
```suggestion
- Packages SHOULD use the [cross-database macros](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros) built into dbt Core where available, such as `{{ except() }}` and `{{ type_string() }}`.
```

Author: Went back and forth on this and maybe came down too hard 🤔. How about packages intended for use with multiple warehouses MUST...

I like the link to the docs

Author: If there's an edge case that the macros don't cover, then I think the "where available" works as an escape hatch

Reviewer: Love your suggestion to introduce the nuanced language of [RFC 6919](https://datatracker.ietf.org/doc/html/rfc6919) where it is useful! 🤩 

News flash: I'm not an expert on the interpretation and application of RFC 2119 or 6919. The following is all informed by my nascent understanding and interpretations.

After reading the ["Guidance in the use of these Imperatives"](https://datatracker.ietf.org/doc/html/rfc2119#section-6) section of RFC 2119, it feels like `SHOULD` is the applicable imperative here. The following stood out to me in particular:
> they must not be used to try to impose a particular method on implementors where the method is not required for interoperability.

A package could ignore some of our "musts" and operate just fine. I didn't check, but I'm guessing there are examples currently in the [Hub](https://hub.getdbt.com/) that would back this up.

But most (not all) of our current usage of `MUST` seems like it should be `SHOULD` instead. `SHOULD` appears to be a **really** strong word in the spec! It shouldn't be confused at all with `OPTIONAL`.

So I think we shouldn't feel like we are pandering or compromising on our expectations in any way when we use `SHOULD`. We should feel like we're laying down something strong, confident, and very assertive.

Author: Tragically, 6919 was published on **1 April** 2013 and was never ratified. (See also https://datatracker.ietf.org/doc/html/rfc2324)

Happy to roll with SHOULD here

- Packages SHOULD disambiguate their resource names to avoid clashes with nodes that are likely to already exist in a project. 
    - For example, packages SHOULD NOT provide a model simply called `users`.

### Releases and updates
- Packages' git tags MUST validate against the regex defined in [version.py](/hubcap/version.py).

> - Packages SHOULD follow the guidance of the [Semantic Versioning Specification](https://semver.org/).

Reviewer: I don't foresee the SemVer specification changing anytime soon, but we specifically referring to 2.0.0, I think.

Optional update:
```suggestion
- Packages SHOULD follow the guidance of [Semantic Versioning 2.0.0](https://semver.org/spec/v2.0.0.html).
```
